### PR TITLE
Making ECAL detector based isolation more robust against noise

### PR DESCRIPTION
**Pull request by swagata87**
#### PR description:

Using ECAL PF rechit thresholds for ECAL detector based isolation to make it more robust against noise. With this change, the distributions of ECAL isolation variables (cone size 0.3 and 0.4) should shift a bit towards lower values. The shift sould be more prominent in endcaps. 

#### PR validation:

Checked with run3 workflows.

### CODE REVIEW

**Comment by jfernan2** [comment]
assign ecal-dpg

**Comment by swagata87** [comment]
type egamma

**Comment by jfernan2** [comment]
please test

**Comment by Sam-Harper** [comment]
thank you @swagata87 , this is much appreciated!

**Comment by jfernan2** [comment]
+1

File: RecoEgamma/EgammaPhotonProducers/src/ConversionTrackCandidateProducer.cc [hunk 1]
```diff
@@ -279,6 +282,8 @@ void ConversionTrackCandidateProducer::produce(edm::Event& theEvent, const edm::
     validEndcapSCHandle = false;
   }
 
+  thresholds = &theEventSetup.getData(ecalPFRechitThresholdsToken_);
```
Comment (thomreis) [inline RecoEgamma/EgammaPhotonProducers/src/ConversionTrackCandidateProducer.cc:285]: Why does this need to be a pointer and can not be a constant reference?

File: RecoEgamma/EgammaIsolationAlgos/src/EgammaRecHitIsolation.cc [hunk 2]
```diff
@@ -96,6 +98,13 @@ double EgammaRecHitIsolation::getSum_(const reco::Candidate* emObject, bool retu
           float phiDiff = reco::deltaPhi(phi, phiclus);
           float energy = j->energy();
 
+          float rhThres = 0.0;
+          if (thresholds != nullptr) {
+            rhThres = (*thresholds)[j->detid()];  // access ECAL PFRechit thresholds for noise cleaning
+          }
```
Comment (mmusich) [inline RecoEgamma/EgammaIsolationAlgos/src/EgammaRecHitIsolation.cc:104]:
```suggestion
          float rhThres = (thresholds != nullptr) ? (*thresholds)[j->detid()] : 0.f;
```

is more compact (of course depending on what you decide about https://github.com/cms-sw/cmssw/pull/43535/files#r1423888388)

File: RecoEgamma/EgammaIsolationAlgos/src/EgammaRecHitIsolation.cc [hunk 3]
```diff
@@ -205,6 +216,14 @@ double EgammaRecHitIsolation::getSum_(const reco::SuperCluster* sc, bool returnE
           double phiDiff = reco::deltaPhi(phi, phiclus);
           double energy = j->energy();
 
+          float rhThres = 0.0;
+          if (thresholds != nullptr) {
+            rhThres = (*thresholds)[j->detid()];  // access ECAL PFRechit thresholds for noise cleaning
+          }
```
Comment (mmusich) [inline RecoEgamma/EgammaIsolationAlgos/src/EgammaRecHitIsolation.cc:222]:
```suggestion
          float rhThres = (thresholds != nullptr) ? (*thresholds)[j->detid()] : 0.f;
```

is more compact (of course depending on what you decide about https://github.com/cms-sw/cmssw/pull/43535/files#r1423888388)

**Review by mmusich — COMMENTED** [review]
unrequested feedback from non-reviewer ;)

File: RecoEgamma/EgammaPhotonProducers/src/ConversionTrackCandidateProducer.cc (hunk 1 above)
Comment (swagata87) [reply RecoEgamma/EgammaPhotonProducers/src/ConversionTrackCandidateProducer.cc:285]: it was just easier, as later on we can check if it is nullptr or not, and depending on that we do certain things. 
We can deliberately pass a nullptr in cases where we do not want to apply the PF recHit thresholds at a later stage.

**Comment by swagata87** [comment]
assign egamma-pog

**Comment by jfernan2** [comment]
+1

**Comment by RSalvatico** [comment]
+1

**Comment by thomreis** [comment]
+ecal-dpg

**Comment by swagata87** [comment]
@cms-sw/dqm-l2 do you have any comment?

**Comment by tjavaid** [comment]
please test

**Comment by tjavaid** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**